### ISSUE TITLE
`nix copy` transfers paths in dependency order

### ISSUE THREAD

**Issue by TravisWhitaker** [NixOS/nix#2559]
The `nix copy` command seems to move paths between stores in dependency order. For example, if I have a dependency graph like:

```
A is free.
B is free.
C needs A and B.
D needs C
```

`nix copy` will first move A and B (in parallel, if possible), then C, then D to the destination store. Typically this is good behavior; a concurrent observer should never see D in the store without C, and should never see C in the store without A and B.

However, for very large stores, maintaining this safety property carries a significant performance cost. For example, consider the case of assembling a binary cache of NARs, i.e. a `file://` store. If there are tens of thousands of store paths to move and lots of CPU cores available for assembling NAR files, `nix copy` will "choke" on narrow parts of the dependency graph. There are binary cache workflows for which there will be no concurrent observers, so one doesn't care if, for example, D were to appear in the store before C, or C before A and B.

It'd be nice to have something like `nix copy --any-order` that ignores the dependency order of transferred paths and allows for linear speedups when building binary caches.

**Comment by shlevy** [comment]
In fact we don't even need a special flag here, we could copy in any order and only serialize marking the paths as valid on the other end.

**Comment by TravisWhitaker** [comment]
@shlevy What exactly happens when a path is marked as valid? Is this something that's recorded in the sqlite DB?

**Comment by edolstra** [comment]
Yeah, it's recorded in SQLite (for `LocalStore` at least). The main issue here is that the `Store` abstraction has no interface for adding a path *without* marking it as valid.

**Comment by TravisWhitaker** [comment]
This is still a problem as of Nix 2.3.10.

**Comment by TravisWhitaker** [comment]
This is still a problem as of Nix 2.3.15

**Comment by schickling** [comment]
Still a problem with Nix `2.5.0`

**Comment by domenkozar** [comment]
Note that `cachix push` doesn't have the dependency bottleneck :)

**Comment by edolstra** [comment]
We do have the interface needed to support this now (`Store::addMultipleToStore()`), since that could in theory write a bunch of paths to the store before registering them. Currently `LocalStore` doesn't have a specialization for `addMultipleToStore()` to do that, but that shouldn't be too hard to implement, which should handle the `nix copy` case. Making the build/substitution loop do this would be harder, since `PathSubstitutionGoal` handles only one path at a time.
